gcc/cp/contracts.cc

```diff
@@ -2621,7 +2628,26 @@ apply_preconditions (tree fndecl)
   if (outline_contracts_p (fndecl))
     add_pre_condition_fn_call (fndecl);
   else
-    emit_preconditions (DECL_CONTRACTS (fndecl));
+    {
+      /* We only care about `noexcept()` case. Functions with `throw`
+      exception specification may not work in presence of throwing
+      contract violation handlers.  */
+     /* A noexcept specification (or throw() with -fnothrow-opt).  */
+      if (TYPE_NOEXCEPT_P (TREE_TYPE (current_function_decl)))
+	{
+	  tree r;
+	  location_t spec_location = DECL_SOURCE_LOCATION (
+	      current_function_decl);
+
+	  r = build_stmt (spec_location, MUST_NOT_THROW_EXPR,
```

Comment:
except.cc has a build_must_not_throw_expr. Maybe that can be used instead of build_stmt?

Reply:
I think not. We want a block scope here, and build_must_not_throw_expr doesn't do that. The current code is fine, and just right.